eng/actions/buildkicker/buildkicker.js

```diff
@@ -0,0 +1,336 @@
+
+            let successfulNeeded = checkruns.total_count;
+            console.log(`Found ${successfulNeeded} check run(s)`);
+            // Iterate over each check run and potentially re-run each
+            for (let i = 0; i < checkruns.total_count; i++) {
+                let run = checkruns.check_runs[i];
+                // Get the rerunability state from the EvaluateRerun function; it may return "success", "stop", "wait" or "rerun".
+                let rerunState = EvaluateRerun(run, checkruns);
+                console.log(`Eval for check[${i}]=${run.name}: ${rerunState}`);
+
+                if (rerunState === "stop") {
+                    // Stop code means to abort build kicker, lets bail out
+                    throw new BuildKickerException(`Error: not enough successful runs to retry \`${run.name}\`.`);
+                }
+                else if (rerunState === "rerun") {
+                    // We want to re-run this instance but we need to check the number of allowable re-runs
+
+                    // Fill in "0" reruns if we haven't rerun it before
+                    if (!rerunCounts.has(run.id)) {
+                        console.log(`No reruns for [${run.id}], setting retry count = 0.`);
+                        rerunCounts.set(run.id, 0);
+                    }
+
+                    // Grab the number of reruns that have been done on this run
+                    let reruns = rerunCounts.get(run.id);
+
+                    // If we still have retries remaining, lets do a rerun
+                    if (reruns < retries) {
+                        // Write the new retries done back to the rerunCounts Map
+                        let newRerunCount = reruns + 1;
+                        console.log(`Setting retry count[${run.id}]=${newRerunCount}.`);
+                        rerunCounts.set(run.id, newRerunCount);
+
+                        // Add a comment saying what we're doing
+                        let timestamp = new Date(Date.now());
+                        let newCommentEntry = `- \`${timestamp.toISOString()}\` Retrying \`${run.name}\`, attempt # \`${newRerunCount + 1}\`, \`${retries - newRerunCount}\` retries left.`;
+                        await AppendCommentContent(newCommentEntry);
+
+                        // Make the request to do a re-run
+                        let reqParams = {
+                            owner: repo_owner,
+                            repo: repo_name,
+                            check_run_id: run.id,
+                          };
+                        console.log("Executing rerequestRun on octokit: " + JSON.stringify(reqParams));
+                        await octokit.rest.checks.rerequestRun(reqParams);
+                    }
+                    else {
+                        // If we are out of retries, abort
+                        throw new BuildKickerException(`Error: out of retries for \`${run.name}\`.`);
```

Comment:
So what happens when this throws or errors out? Does it interact with the build state in any way?

Reply:
Built state, no. If we throw an exception here, it is caught by the outer try/catch and attempts to append the error message to the issue_comment on the PR.

It looks like this:

### Build Kicker

Started kicking build, **Build Kicker** may re-request each check-run at most 3 time(s).<details>
<summary>Details</summary>

**Build Kicker** is a github action defined at `.github/workflows/buildkicker.yml` used to automatically retry failed check-runs.

This instance was invoked by `@kelltrick` by adding a comment with the text `/kickbuild`. 

> Retrying tests until they suceed isn't a great engineering practice; so, use this command with caution.

[GitHub Action Details](https://github.com/kelltrick/dotnet-monitor/actions/runs/1695213077)</details>


`2022-01-13T22:52:36.387Z` @kelltrick, **Build Kicker** run failed!

```
Cannot read property '0' of undefined
```